File: dfinity_wallet/lib/ic_api/web/service_api.dart
@JS()
library dfinity_agent.js;

import 'package:dfinity_wallet/data/canister.dart';
import 'package:js/js.dart';
import '../models.dart';
import 'js_utils.dart';

@JS("createServiceApi")
external Promise<ServiceApi> createServiceApi(dynamic identity);

@JS("ServiceApi")
class ServiceApi {
  @JS("acquireICPTs")
  external Promise<void> acquireICPTs(String accountIdentifier, BigInt icpts);

  @JS("getAccount")
  external Promise<dynamic> getAccount();

  @JS("getBalances")
  external Promise<dynamic> getBalances(Object request);

  @JS("createSubAccount")
  external Promise<dynamic> createSubAccount(String name);

  @JS("updateCanisterSettings")
  external Promise<dynamic> updateCanisterSettings(dynamic settings);

  @JS("sendICPTs")
  external Promise<dynamic> sendICPTs(Object request);

  @JS("createNeuron")
  external Promise<void> createNeuron(dynamic request);

  @JS("retryStakeNeuronNotification")
  external Promise<void> retryStakeNeuronNotification(dynamic request);

  @JS("integrationTest")
  external Promise<void> integrationTest();

  @JS("getTransactions")
  external Promise<dynamic> getTransactions(dynamic request);

  @JS("registerHardwareWallet")
  external Promise<void> registerHardwareWallet(String name, dynamic ledgerIdentity);

  @JS("getNeuron")
  external Promise<dynamic> getNeuron(dynamic neuronId);

  @JS("getNeurons")
  external Promise<dynamic> getNeurons();

  @JS("getPendingProposals")
  external Promise<dynamic> getPendingProposals();

  @JS("getProposalInfo")
  external Promise<dynamic> getProposalInfo(dynamic proposalId);

  @JS("listProposals")
  external Promise<dynamic> listProposals(dynamic request);

  @JS("addHotKey")
  external Promise<dynamic> addHotKey(dynamic request);

  @JS("removeHotKey")
  external Promise<dynamic> removeHotKey(dynamic request);

  @JS("startDissolving")
  external Promise<dynamic> startDissolving(dynamic request);

  @JS("stopDissolving")
  external Promise<dynamic> stopDissolving(dynamic request);

  @JS("increaseDissolveDelay")
  external Promise<dynamic> increaseDissolveDelay(dynamic request);

  @JS("follow")
  external Promise<dynamic> follow(dynamic request);

  @JS("registerVote")
  external Promise<dynamic> registerVote(dynamic request);

  @JS("spawn")
  external Promise<dynamic> spawn(dynamic request);

  @JS("split")
  external Promise<dynamic> split(dynamic request);

  @JS("disburse")
  external Promise<dynamic> disburse(dynamic request);

  @JS("disburseToNeuron")
  external Promise<dynamic> disburseToNeuron(dynamic request);

  @JS("makeMotionProposal")
  external Promise<dynamic> makeMotionProposal(dynamic request);

  @JS("createCanister")
  external Promise<dynamic> createCanister(CreateCanisterRequest request);

  @JS("topupCanister")
  external Promise<void> topupCanister(TopupCanisterRequest request);

  @JS("attachCanister")
  external Promise<void> attachCanister(AttachCanisterRequest request);

  @JS("getCanisterDetails")
  external Promise<void> getCanisterDetails(dynamic canisterId);

  @JS("getCanisters")
  external Promise<dynamic> getCanisters();

  @JS("getIcpToCyclesConversionRate")
  external Promise<double> getIcpToCyclesConversionRate();

  @JS("renameSubAccount")
  external Promise<double> renameSubAccount(RenameSubAccountRequest request);

  @JS("detachCanister")
  external Promise<double> detachCanister(DetachCanisterRequest request);
}


@JS()
@anonymous
class IncreaseDissolveDelayRequest {
  external dynamic get neuronId;

  external num get additionalDissolveDelaySeconds;

  external factory IncreaseDissolveDelayRequest(
      {dynamic neuronId, num additionalDissolveDelaySeconds});
}

@JS()
@anonymous
class FollowRequest {
  external dynamic neuronId;
  external int topic;
  external List<dynamic> followees;

  external factory FollowRequest(
      {dynamic neuronId, int topic, List<dynamic> followees});
}

@JS()
@anonymous
class NeuronIdentifierRequest {
  external dynamic neuronId;

  external factory NeuronIdentifierRequest({dynamic neuronId});
}

@JS()
@anonymous
class CreateNeuronRequest {
  external dynamic stake;
  external int? fromSubAccountId;

  external factory CreateNeuronRequest({dynamic stake, int? fromSubAccountId});
}

@JS()
@anonymous
class RetryStakeNeuronNotificationRequest {
  external dynamic blockHeight;
  external dynamic nonce;
  external int? fromSubAccountId;

  external factory RetryStakeNeuronNotificationRequest({dynamic blockHeight, dynamic nonce, int? fromSubAccountId});
}

@JS()
@anonymous
class DisperseNeuronRequest {
  external dynamic neuronId;
  external dynamic amount;
  external String toAccountId;

  external factory DisperseNeuronRequest(
      {dynamic neuronId, dynamic amount, String toAccountId});
}

@JS()
@anonymous
class SendICPTsRequest {
  external dynamic to;
  external dynamic amount;
  external int? fromSubAccountId;

  external factory SendICPTsRequest(
      {dynamic to, dynamic amount, int? fromSubAccountId});
}

@JS()
@anonymous
class RegisterVoteRequest {
  external dynamic neuronId;
  external dynamic proposal;
  external int vote;

  external factory RegisterVoteRequest(
      {dynamic neuronId, dynamic proposal, int vote});
}

@JS()
@anonymous
class SpawnRequest {
  external dynamic neuronId;
  external dynamic newController;

  external factory SpawnRequest({dynamic neuronId, dynamic newController});
}

@JS()
@anonymous
class CreateCanisterRequest {
  external dynamic stake;
  external int? fromSubAccountId;
  external String name;

  external factory CreateCanisterRequest(
      {dynamic stake, int? fromSubAccountId, String name});
}

@JS()
@anonymous
class TopupCanisterRequest {
  external dynamic stake;
  external int? fromSubAccountId;
  external dynamic targetCanisterId;

  external factory TopupCanisterRequest(
      {dynamic stake, int? fromSubAccountId, dynamic targetCanisterId});
}

@JS()
@anonymous
class AttachCanisterRequest {
  external String name;
  external dynamic canisterId;

  external factory AttachCanisterRequest({String name, dynamic canisterId});
}

@JS()
@anonymous
class UpdateSettingsRequest {
  external dynamic canisterId;
  external UpdateCanisterSettings settings;

  external factory UpdateSettingsRequest({dynamic canisterId, UpdateCanisterSettings settings});
}


@JS()
@anonymous
class UpdateCanisterSettings {
  dynamic controller;

  external factory UpdateCanisterSettings({dynamic controller});
}



enum AttachCanisterResult {
  Ok,
  CanisterAlreadyAttached,
  NameAlreadyTaken,
  CanisterLimitExceeded,
}



class CreateCanisterResponse {
  final String? canisterId;
  final Canister? canister;
  final String? errorMessage;
  final CreateCanisterResult result;

  CreateCanisterResponse({required this.result, this.canisterId, this.canister, this.errorMessage});
}

enum CreateCanisterResult {
  Ok,
  FailedToCreateCanister,
  CanisterAlreadyAttached,
  NameAlreadyTaken,
  CanisterLimitExceeded
}


@JS()
@anonymous
class RenameSubAccountRequest {
  external String newName;
  external String accountIdentifier;

  external factory RenameSubAccountRequest({String newName, String accountIdentifier});
}



@JS()
@anonymous
class DetachCanisterRequest {
  external dynamic canisterId;

  external factory DetachCanisterRequest({dynamic canisterId});
}

File: js-agent/src/ServiceApi.ts
import { AnonymousIdentity, HttpAgent, SignIdentity } from "@dfinity/agent";
import { Option } from "./canisters/option";
import governanceBuilder from "./canisters/governance/builder";
import GovernanceService, {
    AddHotKeyRequest,
    DisburseRequest,
    DisburseResponse,
    DisburseToNeuronRequest,
    DisburseToNeuronResponse,
    EmptyResponse,
    FollowRequest,
    IncreaseDissolveDelayRequest,
    ListProposalsRequest,
    ListProposalsResponse,
    MakeMotionProposalRequest,
    MakeNetworkEconomicsProposalRequest,
    MakeProposalResponse,
    MakeRewardNodeProviderProposalRequest,
    MakeSetDefaultFolloweesProposalRequest,
    NeuronInfo,
    RegisterVoteRequest,
    RemoveHotKeyRequest,
    SpawnRequest,
    SpawnResponse,
    SplitRequest,
    StartDissolvingRequest,
    StopDissolvingRequest
} from "./canisters/governance/model";
import { ProposalInfo } from "./canisters/governance/model";
import ledgerBuilder from "./canisters/ledger/builder";
import LedgerService, {
    GetBalancesRequest,
    SendICPTsRequest
} from "./canisters/ledger/model";
import nnsUiBuilder from "./canisters/nnsUI/builder";
import NnsUiService, {
    AccountDetails,
    AttachCanisterRequest,
    AttachCanisterResult,
    CanisterDetails,
    CreateSubAccountResponse,
    DetachCanisterRequest,
    DetachCanisterResponse,
    GetTransactionsRequest,
    GetTransactionsResponse,
    RegisterHardwareWalletRequest,
    RegisterHardwareWalletResponse,
    RemoveHardwareWalletRequest,
    RemoveHardwareWalletResponse,
    RenameSubAccountRequest,
    RenameSubAccountResponse
} from "./canisters/nnsUI/model";
import icManagementBuilder from "./canisters/icManagement/builder";
import ICManagementService, { CanisterDetailsResponse, UpdateSettingsRequest, UpdateSettingsResponse } from "./canisters/icManagement/model";
import createNeuronImpl, { CreateNeuronRequest } from "./canisters/createNeuron";
import { createCanisterImpl, topupCanisterImpl, CreateCanisterRequest, TopupCanisterRequest, CreateCanisterResponse } from "./canisters/createCanister";
import { AccountIdentifier, BlockHeight, CanisterId, E8s, NeuronId } from "./canisters/common/types";
import { LedgerIdentity } from "@dfinity/identity-ledgerhq";
import { principalToAccountIdentifier } from "./canisters/converter";
import { HOST } from "./canisters/constants";

export default class ServiceApi {
    private readonly ledgerService: LedgerService;
    private readonly nnsUiService: NnsUiService;
    private readonly governanceService: GovernanceService;
    private readonly icManagementService: ICManagementService;
    private readonly identity: SignIdentity;

    constructor(identity: SignIdentity) {
        const agent = new HttpAgent({
            host: HOST,
            identity
        });
        this.ledgerService = ledgerBuilder(agent, identity);
        this.nnsUiService = nnsUiBuilder(agent);
        this.governanceService = governanceBuilder(agent, identity);
        this.icManagementService = icManagementBuilder(agent);
        this.identity = identity;
    }

    /* 
        ACCOUNTS
    */

    public createSubAccount = (name: string) : Promise<CreateSubAccountResponse> => {
        return this.nnsUiService.createSubAccount(name);
    }

    public renameSubAccount = (request: RenameSubAccountRequest) : Promise<RenameSubAccountResponse> => {
        return this.nnsUiService.renameSubAccount(request);
    }

    public registerHardwareWallet = (name: string, identity: LedgerIdentity) : Promise<RegisterHardwareWalletResponse> => {
        const accountIdentifier = principalToAccountIdentifier(identity.getPrincipal());
        const request: RegisterHardwareWalletRequest = {
            name,
            accountIdentifier
        };
        return this.nnsUiService.registerHardwareWallet(request);
    }

    public removeHardwareWallet = (request: RemoveHardwareWalletRequest) : Promise<RemoveHardwareWalletResponse> => {
        return this.nnsUiService.removeHardwareWallet(request);
    }

    public getAccount = async () : Promise<AccountDetails> => {
        const response = await this.nnsUiService.getAccount();
        if ("Ok" in response) {
            return response.Ok;
        } else {
            const accountIdentifier = await this.nnsUiService.addAccount();
            return {
                accountIdentifier,
                subAccounts: [],
                hardwareWalletAccounts: []
            };
        }
    }

    public getBalances = (request: GetBalancesRequest) : Promise<Record<AccountIdentifier, E8s>> => {
        return this.ledgerService.getBalances(request);
    }

    public getTransactions = (request: GetTransactionsRequest) : Promise<GetTransactionsResponse> => {
        return this.nnsUiService.getTransactions(request);
    }

    public sendICPTs = (request: SendICPTsRequest): Promise<BlockHeight> => {
        return this.ledgerService.sendICPTs(request);
    }

    /* 
        GOVERNANCE
    */

    public getNeuron = (neuronId: NeuronId) : Promise<Option<NeuronInfo>> => {
        return this.governanceService.getNeuron(neuronId);
    }

    public getNeurons = () : Promise<Array<NeuronInfo>> => {
        return this.governanceService.getNeurons();
    }

    public getPendingProposals = (): Promise<Array<ProposalInfo>> => {
        return this.governanceService.getPendingProposals();
    }

    public getProposalInfo = (proposalId: bigint) : Promise<Option<ProposalInfo>> => {
        return this.governanceService.getProposalInfo(proposalId);
    }

    public listProposals = (request: ListProposalsRequest) : Promise<ListProposalsResponse> => {
        return this.governanceService.listProposals(request);
    }

    public addHotKey = (request: AddHotKeyRequest) : Promise<EmptyResponse> => {
        return this.governanceService.addHotKey(request);
    }

    public removeHotKey = (request: RemoveHotKeyRequest) : Promise<EmptyResponse> => {
        return this.governanceService.removeHotKey(request);
    }

    public startDissolving = (request: StartDissolvingRequest) : Promise<EmptyResponse> => {
        return this.governanceService.startDissolving(request);
    }

    public stopDissolving = (request: StopDissolvingRequest) : Promise<EmptyResponse> => {
        return this.governanceService.stopDissolving(request);
    }

    public increaseDissolveDelay = (request: IncreaseDissolveDelayRequest) : Promise<EmptyResponse> => {
        return this.governanceService.increaseDissolveDelay(request);
    }

    public follow = (request: FollowRequest) : Promise<EmptyResponse> => {
        return this.governanceService.follow(request);
    }

    public registerVote = (request: RegisterVoteRequest) : Promise<EmptyResponse> => {
        return this.governanceService.registerVote(request);
    }

    public spawn = (request: SpawnRequest) : Promise<SpawnResponse> => {
        return this.governanceService.spawn(request);
    }

    public split = (request: SplitRequest) : Promise<EmptyResponse> => {
        return this.governanceService.split(request);
    }

    public disburse = (request: DisburseRequest) : Promise<DisburseResponse> => {
        return this.governanceService.disburse(request);
    }

    public disburseToNeuron = (request: DisburseToNeuronRequest) : Promise<DisburseToNeuronResponse> => {
        return this.governanceService.disburseToNeuron(request);
    }

    public makeMotionProposal = (request: MakeMotionProposalRequest) : Promise<MakeProposalResponse> => {
        return this.governanceService.makeMotionProposal(request);
    }

    public makeNetworkEconomicsProposal = (request: MakeNetworkEconomicsProposalRequest) : Promise<MakeProposalResponse> => {
        return this.governanceService.makeNetworkEconomicsProposal(request);
    }

    public makeRewardNodeProviderProposal = (request: MakeRewardNodeProviderProposalRequest) : Promise<MakeProposalResponse> => {
        return this.governanceService.makeRewardNodeProviderProposal(request);
    }

    public makeSetDefaultFolloweesProposal = (request: MakeSetDefaultFolloweesProposalRequest) : Promise<MakeProposalResponse> => {
        return this.governanceService.makeSetDefaultFolloweesProposal(request);
    }

    public createNeuron = (request: CreateNeuronRequest) : Promise<NeuronId> => {
        return createNeuronImpl(
            this.identity, 
            this.ledgerService, 
            request);
    }

    /*
        CANISTERS
    */

    public createCanister = (request: CreateCanisterRequest) : Promise<CreateCanisterResponse> => {
        return createCanisterImpl(
            this.identity.getPrincipal(), 
            this.ledgerService, 
            this.nnsUiService,
            request);
    }

    public topupCanister = (request: TopupCanisterRequest) : Promise<boolean> => {
        return topupCanisterImpl(
            this.ledgerService, 
            request);
    }

    public attachCanister = (request: AttachCanisterRequest) : Promise<AttachCanisterResult> => {
        return this.nnsUiService.attachCanister(request);
    }

    public detachCanister = (request: DetachCanisterRequest) : Promise<DetachCanisterResponse> => {
        return this.nnsUiService.detachCanister(request);
    }

    public getCanisters = (): Promise<Array<CanisterDetails>> => {
        return this.nnsUiService.getCanisters();
    }

    public getCanisterDetails = (canisterId: CanisterId): Promise<CanisterDetailsResponse> => {
        return this.icManagementService.getCanisterDetails(canisterId);
    }

    public updateCanisterSettings = (request: UpdateSettingsRequest): Promise<UpdateSettingsResponse> => {
        return this.icManagementService.updateSettings(request);
    }    

    public getIcpToCyclesConversionRate = (): Promise<bigint> => {
        return this.nnsUiService.getIcpToCyclesConversionRate();
    }
}
